### DISCUSSION TITLE
ActorLogic type that takes an Event type in Typescript

### DISCUSSION THREAD

**Discussion by PaulHax** [statelyai/xstate#4647] [Q&A]
I'm wondering how to type an ActorLogic that takes a specific event?  The ActorLogic should also be able to take other unknown events. 
```ts
type MyEvent = { type: 'myEvent' };
type MyActor =  ActorLogic<any, MyEvent | AnyEventObject, any, any>;
function sendMyEvent(myEventRecevingActor: MyActor) {
  myEventRecevingActor.send({ type: 'myEvent' });
  // myEventRecevingActor.send({ type: 'otherEvent' }); // Idealy this would be an error.  
  // It should be OK to pass an actor with other events to this fucntion, but not OK to send those events in this fucntion.
}

const actorThatAlsoTakesOtherEvent = createActor( machineTakingMyEventAndOtherEvent).start()
sendMyEvent(actorThatAlsoTakesOtherEvent) // <-- Wish this worked but getting Typscript error:
// Error:  AnyEventObject is not assignable to OtherEvent
```

Ideally, a `MyActor` typed variable could be assigned an actor that takes events other than `MyEvent`, but would error with
`myActor.send({ type: 'otherEvent' })`

**Comment by davidkpiano** [comment]
I'm not sure if this is currently possible (cc. @Andarist)

**Comment by Andarist** [comment]
```ts
import { setup, createActor, Snapshot, ActorRef } from "xstate"; // types: 5.9.1

type MyEvent = { type: "myEvent" };
type MyActor = ActorRef<Snapshot<unknown>, MyEvent, any>;
function sendMyEvent(myEventRecevingActor: MyActor) {
  myEventRecevingActor.send({ type: "myEvent" });
  // @ts-expect-error
  myEventRecevingActor.send({ type: "otherEvent" });
}

const machineTakingMyEventAndOtherEvent = setup({
  types: {} as {
    events:
      | {
          type: "myEvent";
        }
      | {
          type: "yetAnotherEvent";
        };
  },
}).createMachine({});

const actorThatAlsoTakesOtherEvent = createActor(
  machineTakingMyEventAndOtherEvent,
).start();

// ok
sendMyEvent(actorThatAlsoTakesOtherEvent);
```

[TS playground](https://www.typescriptlang.org/play?ssl=28&ssc=43&pln=1&pc=1#code/JYWwDg9gTgLgBAbzgZwKYwK5gDRwMZSoCGMqAgnjNLgMoB2RYyAFhDLhVVAEqoBmcAL5w+UCCDgAiAB7IYJVJIDccAPSq4MAJ5hUyAFxwArADoAnCYCMAKGvbdcALJaAogDdUdeAF5EmnaiGkiCuHl6SQkp2AU5anNBwvvE8-AA89IwsbKkYdADWdBAA7nQAfLjO7p7scER0WqVRfLmUwBB0KJ4AJpVhMAAUIVVevHiobsB0AObJhs7JAJSI1nBwQ32j45MzlNAmaHRd-Uj2gVLr1RGCC1Gr6nAAAjDIALSo0rqUb1BiUCtroWqmwm02S+26x38uiCbGYqCgwxgVxu1kEtjw7TkayIeGYk1QABUiHltr1qmRDgB5GBwhF9RKdTBgY7-U4GRDCIjIZarVbjaoGf68uAAHx5wolpyCF3CtwlqzR8rFCCF8qhZ0kWnQFNh8MRylVwsEcsE2FRCxMBGIpEcOLxdFQx2uUWsGLoWJxXAJzBIZAANsgIES8npqbTEQyrQpkv1-iA7fjg6TAV4KV0w3q+maLXIiLB+ijrPcIHlrAceimBp7oN7fQGg8TQzTM9UUUA)
